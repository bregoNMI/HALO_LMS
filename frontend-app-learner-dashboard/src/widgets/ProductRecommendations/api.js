import { get, stringifyUrl } from 'data/services/lms/utils';
import urls from 'data/services/lms/urls';

export const crossProductAndAmplitudeRecommendationsUrl = (courseId) => `${urls.getApiUrl()}/edx_recommendations/learner_dashboard/cross_product/${courseId}/`;
export const amplitudeRecommendationsUrl = () => `${urls.getApiUrl()}/edx_recommendations/learner_dashboard/amplitude/v2/`;
export const recommendationsContextUrl = () => `${urls.getApiUrl()}/edx_recommendations/learner_dashboard/recommendations_context/`;

const fetchRecommendationsContext = () => get(stringifyUrl(recommendationsContextUrl()));

const fetchCrossProductRecommendations = (courseId) => (
  get(stringifyUrl(crossProductAndAmplitudeRecommendationsUrl(courseId)))
);
const fetchAmplitudeRecommendations = () => get(stringifyUrl(amplitudeRecommendationsUrl()));

export default {
  fetchCrossProductRecommendations,
  fetchAmplitudeRecommendations,
  fetchRecommendationsContext,
};
