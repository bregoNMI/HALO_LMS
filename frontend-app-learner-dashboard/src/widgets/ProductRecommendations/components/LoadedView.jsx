import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Container } from '@openedx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import messages from '../messages';
import { courseShape, courseTypeToProductTypeMap } from '../utils';
import ProductCardContainer from './ProductCardContainer';

const LoadedView = ({ crossProductCourses, openCourses }) => {
  const { formatMessage } = useIntl();
  const includesCrossProductTypes = crossProductCourses.length === 2;

  const finalProductList = useMemo(() => {
    if (includesCrossProductTypes) {
      const openCourseList = openCourses.slice(0, 2);
      return crossProductCourses.concat(openCourseList);
    }
    return openCourses;
  }, [crossProductCourses, openCourses, includesCrossProductTypes]);

  const courseTypes = [...new Set(finalProductList.map((item) => courseTypeToProductTypeMap[item.courseType]))];

  return (
    <Container
      size="lg"
      className="recommendations-container pt-sm-5 pt-4.5 pb-2 pb-sm-4.5"
    >
      <h2>
        {formatMessage(messages.recommendationsHeading)}
      </h2>
      <ProductCardContainer finalProductList={finalProductList} courseTypes={courseTypes} />
    </Container>
  );
};

LoadedView.propTypes = {
  crossProductCourses: PropTypes.arrayOf(
    PropTypes.shape(courseShape),
  ).isRequired,
  openCourses: PropTypes.arrayOf(
    PropTypes.shape(courseShape),
  ).isRequired,
};

export default LoadedView;
