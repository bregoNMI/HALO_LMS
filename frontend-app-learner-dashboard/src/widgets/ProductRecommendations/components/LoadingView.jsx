import React from 'react';
import { Skeleton } from '@openedx/paragon';

export const LoadingView = () => (
  <Skeleton height={100} />
);

export default LoadingView;
