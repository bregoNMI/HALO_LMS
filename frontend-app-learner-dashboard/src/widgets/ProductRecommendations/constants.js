export const bootCamp = 'Boot Camp';
export const executiveEducation = 'Executive Education';
export const course = 'Course';

export const control = 'control';
export const treatment = 'treatment';
export const noExperiment = 'no experiment';
