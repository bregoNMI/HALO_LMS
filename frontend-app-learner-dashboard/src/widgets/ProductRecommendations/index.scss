@import "@openedx/paragon/scss/core/core";

$horizontal-card-gap: 20px;
$vertical-card-gap: 24px;

.base-card {
  height: 332px;
  width: 270px !important;

  .pgn__card-image-cap {
    height: 104px;
    object: {
      fit: cover;
      position: top center;
    }
  }

  .pgn__card-logo-cap {
    bottom: -1.5rem;
    object: {
      fit: scale-down;
      position: center center;
    }
  }

  .product-card-title {
    font: {
      size: 1.125rem;
    }

    line-height: 24px ;
  }

  .product-card-subtitle {
    font: {
      size: 0.875rem; 
    }

    line-height: 24px;
  }

  .product-badge {
    bottom: 2.75rem;
  }
}

.product-card-container {
  gap: $vertical-card-gap $horizontal-card-gap;
  margin: 0 (-$horizontal-card-gap);
  padding: 1rem $horizontal-card-gap;

  .course-subcontainer {
    gap: $vertical-card-gap $horizontal-card-gap;
  }

  @include media-breakpoint-down(lg) {
    overflow-x: scroll;
  }
}

// Workaround for giving the sub-container a greyish background that stretches to the full width of the browser window
// while being placed within the boundaries of the parent container's dimensions
.recommendations-container::before {
  content: "";
  position: absolute;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
  width: 100vw;
  height: 100%;
  background-color: $light-200;
  z-index: -1;
}
