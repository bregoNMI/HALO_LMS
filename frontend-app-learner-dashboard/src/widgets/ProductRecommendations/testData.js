export const getCoursesWithType = (courseTypes, parameters = true) => {
  const courses = [];
  const marketingUrl = parameters
    ? 'https://www.edx.org/course/some-course?utm_source=source'
    : 'https://www.edx.org/course/some-course';

  courseTypes.forEach((type) => {
    courses.push({
      title: 'Introduction to Computer Science',
      courseRunKey: 'course-v1:Test+Course+2022T2',
      marketingUrl,
      courseType: type,
      image: {
        src: 'https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg',
      },
      owners: [
        {
          key: 'HarvardX',
          name: 'Harvard University',
          logoImageUrl: 'http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png',
        },
      ],
    });
  });

  return courses;
};

export const mockFooterRecommendationsHook = {
  default: { isExperimentActive: false, inRecommendationsVariant: true },
  activeControl: { isExperimentActive: true, inRecommendationsVariant: false },
  activeTreatment: { isExperimentActive: true, inRecommendationsVariant: true },
};

export const mockCrossProductCourses = getCoursesWithType(['executive-education-2u', 'bootcamp-2u']);
export const mockOpenCourses = getCoursesWithType(['verified-audit', 'audit', 'verified', 'course']);
export const mockFallbackOpenCourse = getCoursesWithType(['course'], false);

export const mockCrossProductResponse = {
  crossProductCourses: mockCrossProductCourses,
  amplitudeCourses: mockOpenCourses,
};

export const mockAmplitudeResponse = {
  amplitudeCourses: mockOpenCourses,
};
