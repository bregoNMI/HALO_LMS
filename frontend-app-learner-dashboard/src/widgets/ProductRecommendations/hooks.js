import { useState, useEffect } from 'react';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';

import { RequestStates, RequestKeys } from 'data/constants/requests';
import { StrictDict } from 'utils';
import { reduxHooks } from 'hooks';
import { SortKeys } from 'data/constants/app';
import { useWindowSize, breakpoints } from '@openedx/paragon';
import { useExperimentContext } from 'ExperimentContext';
import { control, treatment, noExperiment } from './constants';
import { activateProductRecommendationsExperiment, trackProductRecommendationsViewed } from './optimizelyExperiment';
import { recommendationsViewed } from './track';

import api from './api';
import * as module from './hooks';

export const state = StrictDict({
  requestState: (val) => useState(val), // eslint-disable-line
  data: (val) => useState(val), // eslint-disable-line
});

export const useIsMobile = () => {
  const { width } = useWindowSize();
  return width < breakpoints.small.minWidth;
};

export const useMostRecentCourseRunKey = () => {
  const mostRecentCourseRunKey = reduxHooks.useCurrentCourseList({
    sortBy: SortKeys.enrolled,
    filters: [],
    pageSize: 0,
  }).visible[0]?.courseRun?.courseId;

  return mostRecentCourseRunKey;
};

export const useActivateRecommendationsExperiment = () => {
  const enterpriseDashboardData = reduxHooks.useEnterpriseDashboardData();
  const hasRequestCompleted = reduxHooks.useRequestIsCompleted(RequestKeys.initialize);
  const mostRecentCourseRunKey = module.useMostRecentCourseRunKey();
  const userId = getAuthenticatedUser().userId.toString();

  const {
    experiment: { isExperimentActive },
    setExperiment,
    isMobile,
    countryCode,
  } = useExperimentContext();

  useEffect(() => {
    if (!isExperimentActive && countryCode !== null) {
      const activateExperiment = () => {
        const userAttributes = {
          is_mobile_user: isMobile,
          is_enterprise_user: !!enterpriseDashboardData,
          location: countryCode ? countryCode.toLowerCase() : '',
        };
        const experiment = activateProductRecommendationsExperiment(userId, userAttributes);

        setExperiment((prev) => ({
          ...prev,
          isExperimentActive: true,
          inRecommendationsVariant: experiment.inExperimentVariant,
        }));

        return experiment;
      };

      const sendViewedEvent = () => {
        trackProductRecommendationsViewed(userId);
        recommendationsViewed(true, control, mostRecentCourseRunKey);
      };

      if (hasRequestCompleted) {
        const { experimentActivated, inExperimentVariant } = activateExperiment();

        if (experimentActivated && !inExperimentVariant) {
          sendViewedEvent();
        }
      }
    }
  /* eslint-disable */
  }, [isExperimentActive, countryCode])
};

export const useShowRecommendationsFooter = () => {
  const { experiment } = useExperimentContext();

  return experiment;
};

export const useFetchRecommendations = (setRequestState, setData) => {
  const courseRunKey = module.useMostRecentCourseRunKey();

  useEffect(() => {
    let isMounted = true;

    const handleSuccess = (response) => {
      if (isMounted) {
        setData(response.data);
        setRequestState(RequestStates.completed);
      }
    };

    const handleError = () => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    };

    if (courseRunKey) {
      api
        .fetchCrossProductRecommendations(courseRunKey)
        .then(handleSuccess)
        .catch(handleError);
    } else {
      api
        .fetchAmplitudeRecommendations()
        .then(handleSuccess)
        .catch(handleError);
    }
    return () => {
      isMounted = false;
    };
    /* eslint-disable */
  }, []);
};

export const useSendViewedEvents = (requestState, data) => {
  const mostRecentCourseRunKey = module.useMostRecentCourseRunKey();
  const userId = getAuthenticatedUser().userId.toString();

  useEffect(() => {
    if (requestState === RequestStates.completed) {
      if (data.crossProductCourses?.length === 2) {
        trackProductRecommendationsViewed(userId);
        recommendationsViewed(false, treatment, mostRecentCourseRunKey);
      } else {
        trackProductRecommendationsViewed(userId);
        recommendationsViewed(true, noExperiment, mostRecentCourseRunKey);
      }
    }
  }, [data, requestState])
}

export const useProductRecommendationsData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});

  module.useFetchRecommendations(setRequestState, setData);
  module.useSendViewedEvents(requestState, data);

  return {
    productRecommendations: data,
    isLoading: requestState === RequestStates.pending,
    isLoaded: requestState === RequestStates.completed,
    hasFailed: requestState === RequestStates.failed 
  };
};

export default { useProductRecommendationsData, useShowRecommendationsFooter, useIsMobile, useActivateRecommendationsExperiment };
