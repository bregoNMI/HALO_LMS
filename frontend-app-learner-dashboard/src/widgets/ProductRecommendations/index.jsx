import React from 'react';
import './index.scss';
import { reduxHooks } from 'hooks';
import NoCoursesView from 'containers/CoursesPanel/NoCoursesView';
import LoadingView from './components/LoadingView';
import LoadedView from './components/LoadedView';
import hooks from './hooks';

const ProductRecommendations = () => {
  const checkEmptyResponse = (obj) => {
    const values = Object.values(obj);
    const result = values.filter((item) => item.length === 0);
    return result.length === values.length;
  };

  const { productRecommendations, isLoading, isLoaded } = hooks.useProductRecommendationsData();
  const isMobile = hooks.useIsMobile();
  const hasCourses = reduxHooks.useHasCourses();
  const shouldShowPlaceholder = checkEmptyResponse(productRecommendations);

  if (isLoading && !isMobile) {
    return <LoadingView />;
  }

  if (isLoaded && !isMobile && !shouldShowPlaceholder) {
    return (
      <LoadedView
        openCourses={productRecommendations.amplitudeCourses}
        crossProductCourses={productRecommendations.crossProductCourses || []}
      />
    );
  }

  if (isLoaded && hasCourses && !isMobile && shouldShowPlaceholder) {
    return <NoCoursesView />;
  }
  return null;
};

export default ProductRecommendations;
