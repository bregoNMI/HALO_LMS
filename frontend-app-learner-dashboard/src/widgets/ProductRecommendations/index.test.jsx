import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { reduxHooks } from 'hooks';
import hooks from './hooks';
import ProductRecommendations from './index';
import LoadingView from './components/LoadingView';
import LoadedView from './components/LoadedView';
import NoCoursesView from '../../containers/CoursesPanel/NoCoursesView';
import { mockCrossProductResponse, mockAmplitudeResponse } from './testData';

jest.mock('./hooks', () => ({
  useProductRecommendationsData: jest.fn(),
  useIsMobile: jest.fn(),
}));

jest.mock('hooks', () => ({
  reduxHooks: {
    useHasCourses: jest.fn(),
  },
}));

jest.mock('./components/LoadingView', () => 'LoadingView');
jest.mock('./components/LoadedView', () => 'LoadedView');
jest.mock('containers/CoursesPanel/NoCoursesView', () => 'NoCoursesView');

describe('ProductRecommendations', () => {
  const defaultValues = {
    productRecommendations: {},
    isLoading: false,
    isLoaded: false,
    hasFailed: false,
  };

  const successfulLoadValues = {
    ...defaultValues,
    isLoaded: true,
    productRecommendations: mockCrossProductResponse,
  };

  beforeEach(() => reduxHooks.useHasCourses.mockReturnValueOnce(true));

  it('matches snapshot', () => {
    hooks.useIsMobile.mockReturnValueOnce(false);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfulLoadValues,
    });

    expect(shallow(<ProductRecommendations />).snapshot).toMatchSnapshot();
  });

  it('renders the LoadingView if the request is pending', () => {
    hooks.useIsMobile.mockReturnValueOnce(false);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...defaultValues,
      isLoading: true,
    });

    expect({ ...shallow(<ProductRecommendations />).shallowWrapper, children: expect.any(Array) }).toMatchObject(
      shallow(<LoadingView />),
    );
  });
  it('renders nothing if the request has failed', () => {
    hooks.useIsMobile.mockReturnValueOnce(false);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...defaultValues,
      hasFailed: true,
    });

    const wrapper = shallow(<ProductRecommendations />);

    expect(wrapper.shallowWrapper).toBeNull();
  });
  it('renders nothing if the user is on the mobile view', () => {
    hooks.useIsMobile.mockReturnValueOnce(true);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfulLoadValues,
    });

    const wrapper = shallow(<ProductRecommendations />);

    expect(wrapper.shallowWrapper).toBeNull();
  });

  it('renders NoCoursesView if the request is loaded, user has courses, and the response is empty', () => {
    hooks.useIsMobile.mockReturnValueOnce(false);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfulLoadValues,
      productRecommendations: {
        amplitudeCourses: [],
        crossProductCourses: [],
      },
    });

    expect({ ...shallow(<ProductRecommendations />).shallowWrapper, children: expect.any(Array) }).toMatchObject(
      shallow(<NoCoursesView />),
    );
  });

  describe('LoadedView', () => {
    it('renders with cross product data if the request completed and the user has courses', () => {
      hooks.useIsMobile.mockReturnValueOnce(false);
      hooks.useProductRecommendationsData.mockReturnValueOnce({
        ...successfulLoadValues,
      });

      expect({ ...shallow(<ProductRecommendations />).shallowWrapper, children: expect.any(Array) }).toMatchObject(
        shallow(
          <LoadedView
            openCourses={mockCrossProductResponse.amplitudeCourses}
            crossProductCourses={mockCrossProductResponse.crossProductCourses}
          />,
        ),
      );
    });

    it('renders the LoadedView with Amplitude course data if the request completed', () => {
      hooks.useIsMobile.mockReturnValueOnce(false);
      hooks.useProductRecommendationsData.mockReturnValueOnce({
        ...successfulLoadValues,
        productRecommendations: mockAmplitudeResponse,
      });

      expect({ ...shallow(<ProductRecommendations />).shallowWrapper, children: expect.any(Array) }).toMatchObject(
        shallow(
          <LoadedView
            openCourses={mockCrossProductResponse.amplitudeCourses}
            crossProductCourses={[]}
          />,
        ),
      );
    });
  });
});
