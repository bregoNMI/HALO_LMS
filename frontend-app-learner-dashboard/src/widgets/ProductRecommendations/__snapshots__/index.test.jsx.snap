// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRecommendations matches snapshot 1`] = `
<LoadedView
  crossProductCourses={
    Array [
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "executive-education-2u",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "bootcamp-2u",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
    ]
  }
  openCourses={
    Array [
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "verified-audit",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "audit",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "verified",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseRunKey": "course-v1:Test+Course+2022T2",
        "courseType": "course",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "marketingUrl": "https://www.edx.org/course/some-course?utm_source=source",
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "title": "Introduction to Computer Science",
      },
    ]
  }
/>
`;
